applications/IgaApplication/custom_elements/membrane_element.h

```diff
@@ -0,0 +1,490 @@
+#if !defined(KRATOS_MEMBRANE_ELEMENT_H_INCLUDED )
+#define  KRATOS_MEMBRANE_ELEMENT_H_INCLUDED
+
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/element.h"
+#include "utilities/math_utils.h"
+
+// Application includes
+#include "iga_application_variables.h"
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+/// Short class definition.
+/** Kirchhoff Membrane. 
+*/
+class MembraneElement
+    : public Element
+{
+protected:
+
+    /// Internal variables used for metric transformation
+    struct KinematicVariables
+    {
+        // covariant metric
+        array_1d<double, 3> a_ab_covariant;
+
+        //base vector 1
+        array_1d<double, 3> a1;
+        //base vector 2
+        array_1d<double, 3> a2;
+        //base vector 3 normalized
+        array_1d<double, 3> a3;
+        //not-normalized base vector 3
+        array_1d<double, 3> a3_tilde;
+
+        //differential area
+        double dA;
+
+        /**
+        * The default constructor
+        * @param Dimension: The size of working space dimension
+        */
+        KinematicVariables(SizeType Dimension)
+        {
+            noalias(a_ab_covariant) = ZeroVector(Dimension);
+
+            noalias(a1) = ZeroVector(Dimension);
+            noalias(a2) = ZeroVector(Dimension);
+            noalias(a3) = ZeroVector(Dimension);
+
+            noalias(a3_tilde) = ZeroVector(Dimension);
+
+            dA = 1.0;
+        }
+    };
+
+    /**
+    * Internal variables used in the constitutive equations
+    */
+    struct ConstitutiveVariables
+    {
+        Vector StrainVector;
+        Vector StressVector;
+        Matrix ConstitutiveMatrix;
+
+        /**
+        * @param StrainSize: The size of the strain vector in Voigt notation
+        */
+        ConstitutiveVariables(SizeType StrainSize)
+        {
+            StrainVector = ZeroVector(StrainSize);
+            StressVector = ZeroVector(StrainSize);
+            ConstitutiveMatrix = ZeroMatrix(StrainSize, StrainSize);
+        }
+    };
+
+    /**
+    * Internal variables used in the constitutive equations
+    */
+    struct SecondVariations
+    {
+        Matrix B11;
+        Matrix B22;
+        Matrix B12;
+
+        /**
+        * The default constructor
+        * @param StrainSize: The size of the strain vector in Voigt notation
+        */
+        SecondVariations(const int& mat_size)
+        {
+            B11 = ZeroMatrix(mat_size, mat_size);
+            B22 = ZeroMatrix(mat_size, mat_size);
+            B12 = ZeroMatrix(mat_size, mat_size);
+        }
+    };
+
+    struct PrestresstransVariables
+    {
+        Matrix Tpre;
+        
+        PrestresstransVariables(const unsigned int& Dimension)
+        {
+           Matrix Tpre = ZeroMatrix(3, 3);
+        }
+    };
+
+    //...is it needed?...
+    struct transPreVariables
```

Comment:
you don't need to define a struct if it is just one variable.

Second, you know the size already, thus you can use array_1d<double, 3> instead of vector

Reply:
anyways, the correct naming would be TransPre...